server/s3_motr_writer.cc

```diff
@@ -724,22 +737,119 @@ void S3MotrWiter::set_up_motr_data_buffers(struct s3_motr_rw_op_context *rw_ctx,
     rw_ctx->data->ov_buf[buf_idx] = ptr_n_len.first;
     rw_ctx->data->ov_vec.v_count[buf_idx] = size_of_each_buf;
 
-    // Here we use actual length to get md5
-    md5crypt.Update((const char *)ptr_n_len.first, len_in_buf);
+    if (!is_s3_write_di_check_enabled) {
+      // Here we use actual length to get md5
+      md5crypt.Update((const char *)ptr_n_len.first, len_in_buf);
+    } else {
+      rw_ctx->pi_bufvec->ov_buf[0] = ptr_n_len.first;
+      rw_ctx->pi_bufvec->ov_vec.v_count[0] = size_of_each_buf;
+      rw_ctx->pi_bufvec->ov_vec.v_nr = 1;
+
+      if (last_index > 0) {
+        rc = s3_motr_api->motr_client_calculate_pi(
+            &rw_ctx->pi, NULL, rw_ctx->pi_bufvec,
+            static_cast<m0_pi_calc_flag>(0),
+            (unsigned char *)rw_ctx->current_digest, NULL);
+        if (rc != 0) {
+          s3_log(
+              S3_LOG_ERROR, request_id,
+              "motr api motr_client_calculate_pi failed with return code %d\n",
+              rc);
+        }
+        assert(rc == 0);
+      } else {
+        // Checksum initialization and also update will happen one after another
+        rc = s3_motr_api->motr_client_calculate_pi(
+            &rw_ctx->pi, NULL, rw_ctx->pi_bufvec, M0_PI_CALC_UNIT_ZERO,
+            (unsigned char *)rw_ctx->current_digest, NULL);
+        if (rc != 0) {
+          s3_log(
+              S3_LOG_ERROR, request_id,
+              "motr api motr_client_calculate_pi failed with return code %d\n",
+              rc);
+        }
+        assert(rc == 0);
+        // save digest of checksum init
+        // md5crypt.save_motr_unit_checksum(rw_ctx->pi.t_pi->prev_digest);
+        md5crypt.save_motr_unit_checksum((unsigned char *)rw_ctx->pi.t_pi);
+      }
+    }
 
     // Init motr buffer attrs.
     rw_ctx->ext->iv_index[buf_idx] = last_index;
     rw_ctx->ext->iv_vec.v_count[buf_idx] = /*data_len*/ size_of_each_buf;
     last_index += /*data_len*/ size_of_each_buf;
 
     /* we don't want any attributes */
-    rw_ctx->attr->ov_vec.v_count[buf_idx] = 0;
+    // rw_ctx->attr->ov_vec.v_count[buf_idx] = 0;
 
     ++buf_idx;
     size_in_current_write += len_in_buf;
+    if (is_s3_write_di_check_enabled) {
+      if (size_in_current_write % motr_unit_size == 0) {
+        struct m0_pi_seed seed;
+        seed.data_unit_offset = saved_last_index;
+        seed.obj_id.f_container = oid_list[0].u_hi;
+        seed.obj_id.f_key = oid_list[0].u_lo;
+        saved_last_index = last_index;
+        assert(chksum_buf_idx <= rw_ctx->motr_checksums_buf_count);
+
+        // memcpy(((struct m0_generic_pi
+        // *)rw_ctx->attr->ov_buf[chksum_buf_idx])->t_pi,
+        // md5crypt.get_prev_unit_checksum(), PI_DIGEST_LENGTH);
+        memcpy(
+            ((struct m0_md5_inc_digest_pi *)((struct m0_generic_pi *)rw_ctx
```

Comment:
Let us convert this code in utility function/common class which can be used for GET as well